### PR TITLE
Add WS to backup/restore profiles

### PR DESCRIPTION
**Pull request by jblievremont**
Since backup/restore components are 1. based on DaoV2 and 2. already medium tested, provided unit tests for backup and restore WS only test the surface checks (WS params, permissions).

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileBackupAction.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    backuper.backup(request.mandatoryParam(PARAM_KEY), new OutputStreamWriter(stream.output(), Charsets.UTF_8));
+  }
+
+  @Override
+  public void define(WebService.NewController controller) {
+    controller.createAction("backup")
+      .setSince("5.2")
+      .setDescription("Backup a quality profile in XML form.")
+      .setHandler(this)
+      .createParam(PARAM_KEY)
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileBackupAction.java]: as discussed the parameters "profileName" and "language" should be supported in addition to "key". For this reason I would rename "key" to "profileKey" (and to be consistent though)

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileBackupAction.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+  public QProfileBackupAction(QProfileBackuper backuper) {
+    this.backuper = backuper;
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    Stream stream = response.stream();
+    stream.setMediaType("text/xml");
+    backuper.backup(request.mandatoryParam(PARAM_KEY), new OutputStreamWriter(stream.output(), Charsets.UTF_8));
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileBackupAction.java]: writer is not closed

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileBackupAction.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  @Override
+  public void define(WebService.NewController controller) {
+    controller.createAction("backup")
+      .setSince("5.2")
+      .setDescription("Backup a quality profile in XML form.")
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileBackupAction.java]: IMO you could mention that backup file can be restored through api/qualityprofiles/restore

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileRestoreAction.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  private final QProfileBackuper backuper;
+
+  public QProfileRestoreAction(QProfileBackuper backuper) {
+    this.backuper = backuper;
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    UserSession.get().checkLoggedIn().checkGlobalPermission(GlobalPermissions.QUALITY_PROFILE_ADMIN);
+    InputStream backup = request.paramAsInputStream(PARAM_BACKUP);
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileRestoreAction.java]: backup is not closed

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileRestoreAction.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    backuper.restore(new InputStreamReader(backup, Charsets.UTF_8), null);
+    response.noContent();
+  }
+
+  @Override
+  public void define(WebService.NewController controller) {
+    controller.createAction("restore")
+      .setSince("5.2")
+      .setDescription("Restore a quality profile using an XML file.")
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileRestoreAction.java]: You should mention the behavior when the profile name already exists. Is it overridden or is backup ignored ? In the latter case, should a different HTTP status be used ?

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileRestoreAction.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      .setPost(true)
+      .setHandler(this)
+      .createParam(PARAM_BACKUP)
+      .setDescription("A profile backup file in XML format, as generated by api/qualityprofiles/backup " +
+        "or the former api/profiles/backup.");
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileRestoreAction.java]: no need to mention this deprecated WS

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileRestoreActionTest.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  // TODO Replace with proper DbTester + EsTester medium test once DaoV2 is removed
+  @Mock
+  private QProfileBackuper backuper;
+
+  private WsTester tester;
+
+  @Before
+  public void setUp() throws Exception {
+    tester = new WsTester(new QProfilesWs(
+      mock(RuleActivationActions.class),
```
Comment (simonbrandhof) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileRestoreActionTest.java]: why the 3 first mocks are not BaseQProfileWsAction ?

File: server/sonar-server/src/main/resources/org/sonar/server/qualityprofile/ws/example-projects.json [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+      "selected": false
+    },
+    {
+      "key": "6899e06b-4a34-4966-80f1-073a658bee03",
+      "name": "SonarQube Web Plugin",
+      "selected": false
+    }
+  ],
+
+  "more": false
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/resources/org/sonar/server/qualityprofile/ws/example-projects.json]: I suppose that this is not the correct example. Field "more" should not be used but "total", "p" (page) and "ps" (page size)

Comment (jblievremont) [inline server/sonar-server/src/main/resources/org/sonar/server/qualityprofile/ws/example-projects.json]: See http://jira.codehaus.org/browse/SONAR-6391

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileRestoreActionTest.java (hunk 7 above)
Comment (jblievremont) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileRestoreActionTest.java]: They are not really actions, but classes that define similar actions: activate_rule/deactivate_rule, bulk_activate_rule/bulk_deactivate_rule, add_project/remove_project. But I'll see how I can refactor this.

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileRestoreAction.java (hunk 5 above)
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileRestoreAction.java]: The current behavior is to overwrite the existing profile. I'm adding a warning in the WS description.

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileRestoreAction.java (hunk 6 above)
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileRestoreAction.java]: Well, it felt like a natural way to tell users "don't throw your existing profile backups yet" :)

**Comment by GithubIntegration** [comment]
SonarQube analysis reported 1 new issue:
* 1 major

**Comment by simonbrandhof** [comment]
LGTM

**Pull request closed without merging**